src/meta-srv/src/procedure/region_failover/activate_region.rs

```diff
@@ -0,0 +1,252 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::time::Duration;
+
+use api::v1::meta::mailbox_message::Payload;
+use api::v1::meta::MailboxMessage;
+use async_trait::async_trait;
+use common_meta::instruction::{Instruction, InstructionReply, SimpleReply};
+use common_meta::{DatanodeId, RegionIdent};
+use common_telemetry::debug;
+use serde::{Deserialize, Serialize};
+use snafu::{ensure, OptionExt, ResultExt};
+
+use super::update_metadata::UpdateRegionMetadata;
+use super::{RegionFailoverContext, State};
+use crate::error::{
+    DeserializeFromJsonSnafu, Error, Result, RetryLaterSnafu, SerializeToJsonSnafu,
+    UnexpectedInstructionReplySnafu,
+};
+use crate::service::mailbox::{Channel, MailboxReceiver};
+
+#[derive(Serialize, Deserialize, Debug)]
+pub(super) struct ActivateRegion {
+    candidate: DatanodeId,
+}
+
+impl ActivateRegion {
+    pub(super) fn new(candidate: DatanodeId) -> Self {
+        Self { candidate }
+    }
+
+    async fn send_open_region_message(
+        &self,
+        ctx: &RegionFailoverContext,
+        failed_region: &RegionIdent,
+        timeout: Duration,
+    ) -> Result<MailboxReceiver> {
+        let instruction = Instruction::OpenRegion(failed_region.clone());
+
+        let msg = MailboxMessage {
+            id: 0, // "id" will be set by the mailbox.
+            subject: "Activate Region".to_string(),
+            from: "Metasrv".to_string(),
+            to: format!("Datanode-{}", self.candidate),
+            timestamp_millis: common_time::util::current_time_millis(),
+            payload: Some(Payload::Json(
+                serde_json::to_string(&instruction).with_context(|_| SerializeToJsonSnafu {
+                    input: format!("{instruction:?}"),
+                })?,
+            )),
+        };
+
+        let candidate = Channel::Datanode(self.candidate);
+        ctx.mailbox.send(&candidate, msg, timeout).await
+    }
+
+    async fn handle_response(
+        self,
+        mailbox_receiver: MailboxReceiver,
+        failed_region: &RegionIdent,
+    ) -> Result<Box<dyn State>> {
+        match mailbox_receiver.await? {
+            Ok(msg) => {
+                debug!("Received activate region reply: {msg:?}");
+
+                let Payload::Json(payload) =
+                    msg.payload
+                        .as_ref()
+                        .with_context(|| UnexpectedInstructionReplySnafu {
+                            reply: None,
+                            reason: format!("empty payload, msg: {msg:?}"),
+                        })?;
+                let reply: InstructionReply = serde_json::from_str(payload)
```

Comment:
These json encoding/decoding should be extracted into a helper mod, and use them here. These functions will be used in other places.